Fix some pages with COOP: same-origin header getting a same-site process from cache

zakariaridouh opened the pull request:
#### 54120658172b17a39be914fbbb0fbe6f812597b6
<pre>
Fix some pages with COOP: same-origin header getting a same-site process from cache
<a href="https://bugs.webkit.org/show_bug.cgi?id=293321">https://bugs.webkit.org/show_bug.cgi?id=293321</a>
<a href="https://rdar.apple.com/150643925">rdar://150643925</a>

Reviewed by Chris Dumez.

We have an issue in our process selection system for loading pages,
where if we have a suspended process of a certain site, we sometimes end
up giving that process to a new site that is same domain, but different
origin. If this site also has a Cross-Origin-Opener-Policy: same-origin
header, we end up in a process swapping loop where we keep trying to get
a new process to load in to satisfy our policy, but keep getting handed
a process that we can&apos;t load in.

Fix this by disqualifying COOP + COEP process swaps from eligibility in
participating in process cache selection.

Also, added an API test to verify this fixed logic.

* Source/WebKit/UIProcess/SuspendedPageProxy.cpp:
(WebKit::SuspendedPageProxy::findReusableSuspendedPageProcess):
(WebKit::SuspendedPageProxy::create):
(WebKit::SuspendedPageProxy::SuspendedPageProxy):
* Source/WebKit/UIProcess/SuspendedPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::suspendCurrentPageIfPossible):
(WebKit::WebPageProxy::triggerBrowsingContextGroupSwitchForNavigation):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::processForSite):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessSwapOnNavigation.mm:
((ProcessSwap, MultitabCOOPSwapSameOriginProcessPool)):

Canonical link: <a href="https://commits.webkit.org/295891@main">https://commits.webkit.org/295891@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0f9f28bc81ca2b61c71e3d8e69d836903644db26

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/106451 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/26200 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/16598 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/111653 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/57047 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/108490 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/26867 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/34703 "Built successfully") | [⏳ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/WPE-WK2-Tests-EWS "Waiting to run tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/57047 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/109455 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/21304 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/96059 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/61176 "Found 1 new API test failure: /WPE/TestSSL:/webkit/WebKitWebView/ephemeral-tls-errors (failure)") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/20774 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/14158 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/56488 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/90637 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/14192 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/114513 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/33589 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/24765 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/89921 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/33953 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/92287 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89625 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22875 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/34522 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/12336 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/29198 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/33514 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/38926 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/33260 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/36613 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/34858 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/2cdba2b2f746a482af9279219ebdebe47bb2eecd. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/45664

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/4883b51df82d4b18d2f9e7ebdfa658149927dd3e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/45664

cdumez commented on Source/WebKit/UIProcess/SuspendedPageProxy.cpp:
```diff
@@ -68,7 +69,8 @@ RefPtr<WebProcessProxy> SuspendedPageProxy::findReusableSuspendedPageProcess(Web
             && !process->wasTerminated()
-            && process->hasSameGPUAndNetworkProcessPreferencesAs(pageConfiguration)) {
+            && process->hasSameGPUAndNetworkProcessPreferencesAs(pageConfiguration)
+            && processSwapDisposition != WebCore::ProcessSwapDisposition::COOP) {
```
This needs API testing. We have a lot of testing for COOP/COEP in ProcessSwapOnNavigation.mm

zakariaridouh replied:
Updated PR with an API test & different approach to allow using suspended WebProcesses on COOP process swaps.

cdumez commented on Source/WebKit/UIProcess/SuspendedPageProxy.cpp:
```diff
@@ -58,8 +59,8 @@ static WeakHashSet<SuspendedPageProxy>& allSuspendedPages()
-RefPtr<WebProcessProxy> SuspendedPageProxy::findReusableSuspendedPageProcess(WebProcessPool& processPool, const RegistrableDomain& registrableDomain, WebsiteDataStore& dataStore, WebProcessProxy::LockdownMode lockdownMode, const API::PageConfiguration& pageConfiguration)
-{
+RefPtr<WebProcessProxy> SuspendedPageProxy::findReusableSuspendedPageProcess(WebProcessPool& processPool, const RegistrableDomain& registrableDomain, WebsiteDataStore& dataStore, WebProcessProxy::LockdownMode lockdownMode, const API::PageConfiguration& pageConfiguration, WebCore::ProcessSwapDisposition processSwapDisposition)
+{ // zakr: todo, see if we can just compare the WebCore::SecurityOriginData directly
```
Was this PR meant to be a draft?

zakariaridouh replied:
Nope, accidentally left the comment. Removed!

achristensen07 commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -1221,7 +1223,7 @@ Ref<WebProcessProxy> WebProcessPool::processForSite(WebsiteDataStore& websiteDat
 
         // RefPtr if we have a suspended page for the given registrable domain and use its process if we do, for performance reasons.
-        if (RefPtr process = SuspendedPageProxy::findReusableSuspendedPageProcess(*this, site->domain(), websiteDataStore, lockdownMode, pageConfiguration)) {
+        if (RefPtr process = SuspendedPageProxy::findReusableSuspendedPageProcess(*this, site->domain(), websiteDataStore, lockdownMode, pageConfiguration, WebCore::SecurityOriginData().fromURL(destinationURL), processSwapDisposition)) {
```
(). should be ::

achristensen07 commented on Source/WebKit/UIProcess/WebProcessPool.h:
```diff
@@ -344,7 +345,7 @@ class WebProcessPool final
     void reportWebContentCPUTime(Seconds cpuTime, uint64_t activityState);
 
-    Ref<WebProcessProxy> processForSite(WebsiteDataStore&, const std::optional<WebCore::Site>&, WebProcessProxy::LockdownMode, const API::PageConfiguration&); // Will return an existing one if limit is met or due to caching.
+    Ref<WebProcessProxy> processForSite(WebsiteDataStore&, const std::optional<WebCore::Site>&, WebProcessProxy::LockdownMode, const API::PageConfiguration&, WebCore::ProcessSwapDisposition = WebCore::ProcessSwapDisposition::None, const URL& destinationURL = URL()); // Will return an existing one if limit is met or due to caching.
```
It is confusing to pass in a Site and a URL that should be the same.  We should pass one or the other but not both.  Maybe rename the function too.  processForNavigationToURL?

achristensen07 commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -3180,4 +3180,3 @@ void WebProcessProxy::setResourceMonitorRuleLists(RefPtr<WebCompiledContentRuleL
 #undef MESSAGE_CHECK_COMPLETION
 #undef WEBPROCESSPROXY_RELEASE_LOG
 #undef WEBPROCESSPROXY_RELEASE_LOG_ERROR
-
```
No need to touch this file.

achristensen07 commented on Source/WebKit/UIProcess/WebProcessPool.h:
```diff
@@ -41,6 +41,7 @@
 #include "WebsiteDataStore.h"
 #include <WebCore/CrossSiteNavigationDataTransfer.h>
 #include <WebCore/ProcessIdentifier.h>
+#include <WebCore/ProcessSwapDisposition.h>
```
I think we should not have default values and we should forward declare ProcessSwapDisposition instead of including another header from a header and slowing down the build.  Same elsewhere.

achristensen07 commented on Source/WebKit/UIProcess/SuspendedPageProxy.h:
```diff
@@ -135,6 +137,7 @@ class SuspendedPageProxy final: public IPC::MessageReceiver, public RefCounted<S
     LayerHostingContextID m_contextIDForVisibilityPropagationInGPUProcess { 0 };
 #endif
 #endif
+    WebCore::SecurityOriginData m_securityOriginData;
```
Can this be const?

zakariaridouh replied:
Yes it can, since we only set it when constructing the object. Made it const.

charliewolfe commented on Source/WebKit/UIProcess/SuspendedPageProxy.cpp:
```diff
@@ -58,10 +60,13 @@ static WeakHashSet<SuspendedPageProxy>& allSuspendedPages()
     for (Ref suspendedPage : allSuspendedPages()) {
         Ref process = suspendedPage->process();
+        if (processSwapDisposition == WebCore::ProcessSwapDisposition::COOP && securityOriginData != suspendedPage->m_securityOriginData)
+            return nullptr;
```
This should be `continue`, not `return`

zakariaridouh replied:
Good catch, did all my testing with one suspended page in the cache. Fixed.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a081244b29cef099d55e67c73f86e80fb2142f32. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/45664

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/f829edf7ccebe27664880ec757a4e407433b11cf. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/45664

zakariaridouh commented on Source/WebKit/UIProcess/SuspendedPageProxy.h:
```diff
@@ -59,13 +61,13 @@ class SuspendedPageProxy final: public IPC::MessageReceiver, public RefCounted<S
     WTF_OVERRIDE_DELETE_FOR_CHECKED_PTR(SuspendedPageProxy);
 public:
-    static Ref<SuspendedPageProxy> create(WebPageProxy&, Ref<WebProcessProxy>&&, Ref<WebFrameProxy>&& mainFrame, Ref<BrowsingContextGroup>&&, ShouldDelayClosingUntilFirstLayerFlush);
+    static Ref<SuspendedPageProxy> create(WebPageProxy&, Ref<WebProcessProxy>&&, Ref<WebFrameProxy>&& mainFrame, Ref<BrowsingContextGroup>&&, ShouldDelayClosingUntilFirstLayerFlush, WebCore::SecurityOriginData);
```
`SecurityOriginData` is no longer used in this patch.

zakariaridouh commented on Source/WebKit/UIProcess/SuspendedPageProxy.cpp:
```diff
@@ -107,12 +112,12 @@ static const MessageNameSet& messageNamesToIgnoreWhileSuspended()
 #endif
 
-Ref<SuspendedPageProxy> SuspendedPageProxy::create(WebPageProxy& page, Ref<WebProcessProxy>&& process, Ref<WebFrameProxy>&& mainFrame, Ref<BrowsingContextGroup>&& browsingContextGroup, ShouldDelayClosingUntilFirstLayerFlush shouldDelayClosingUntilFirstLayerFlush)
+Ref<SuspendedPageProxy> SuspendedPageProxy::create(WebPageProxy& page, Ref<WebProcessProxy>&& process, Ref<WebFrameProxy>&& mainFrame, Ref<BrowsingContextGroup>&& browsingContextGroup, ShouldDelayClosingUntilFirstLayerFlush shouldDelayClosingUntilFirstLayerFlush, WebCore::SecurityOriginData securityOriginData)
```
`SecurityOriginData` is no longer used in this patch.

cdumez commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -1211,8 +1212,12 @@ void WebProcessPool::disconnectProcess(WebProcessProxy& process)
+Ref<WebProcessProxy> WebProcessPool::processForSite(WebsiteDataStore& websiteDataStore, const std::optional<Site>& site, WebProcessProxy::LockdownMode lockdownMode, const API::PageConfiguration& pageConfiguration, ProcessSwapDisposition processSwapDisposition)
 {
+    // We don't reuse cached processess because the process cache is per site, whereas COOP swaps are based on origin.
+    if (processSwapDisposition == ProcessSwapDisposition::COOP)
```
Actually, this is not ideal, we do want the `tryTakePrewarmedProcess` logic below. We should totally use a prewarmed process is one is available.

zakariaridouh replied:
Sure, added the check into the conditional right below alongside `site && !site->isEmpty()`

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/0f9f28bc81ca2b61c71e3d8e69d836903644db26. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/45664

zakariaridouh committed:
Fix some pages with COOP: same-origin header getting a same-site process from cache
https://bugs.webkit.org/show_bug.cgi?id=293321
rdar://150643925

Reviewed by Chris Dumez.

We have an issue in our process selection system for loading pages,
where if we have a suspended process of a certain site, we sometimes end
up giving that process to a new site that is same domain, but different
origin. If this site also has a Cross-Origin-Opener-Policy: same-origin
header, we end up in a process swapping loop where we keep trying to get
a new process to load in to satisfy our policy, but keep getting handed
a process that we can't load in.

Fix this by disqualifying COOP + COEP process swaps from eligibility in
participating in process cache selection.

Also, added an API test to verify this fixed logic.

* Source/WebKit/UIProcess/SuspendedPageProxy.cpp:
(WebKit::SuspendedPageProxy::findReusableSuspendedPageProcess):
(WebKit::SuspendedPageProxy::create):
(WebKit::SuspendedPageProxy::SuspendedPageProxy):
* Source/WebKit/UIProcess/SuspendedPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::suspendCurrentPageIfPossible):
(WebKit::WebPageProxy::triggerBrowsingContextGroupSwitchForNavigation):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::processForSite):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessSwapOnNavigation.mm:
((ProcessSwap, MultitabCOOPSwapSameOriginProcessPool)):

Canonical link: https://commits.webkit.org/295891@main
